src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,353 @@
+                                seatNonBid -> seatNonBid.getNonBid().getFirst(),
+                                (existing, replacement) -> existing));
+
+        final List<GreenbidsAdUnit> adUnitsWithBidResponses = imps.stream().map(imp -> createAdUnit(
+                imp, seatsWithBids, seatsWithNonBids)).toList();
+
+        final String auctionId = auctionContextOptional
+                .map(AuctionContext::getBidRequest)
+                .map(BidRequest::getId)
+                .orElse(null);
+
+        final String referrer = auctionContextOptional
+                .map(AuctionContext::getBidRequest)
+                .map(BidRequest::getSite)
+                .map(Site::getPage)
+                .orElse(null);
+
+        return CommonMessage.builder()
+                        .version(greenbidsAnalyticsProperties.getAnalyticsServerVersion())
+                        .auctionId(auctionId)
+                        .referrer(referrer)
+                        .sampling(greenbidsImpExt.getGreenbidsSampling())
+                        .prebidServer(prebidVersionProvider.getNameVersionRecord())
+                        .greenbidsId(greenbidsId)
+                        .pbuid(greenbidsImpExt.getPbuid())
+                        .billingId(billingId)
+                        .adUnits(adUnitsWithBidResponses)
+                        .auctionElapsed(auctionElapsed)
+                        .build();
+    }
+
+    private static SeatNonBid toSeatNonBid(String bidder, BidRejectionTracker bidRejectionTracker) {
+        final List<NonBid> nonBids = bidRejectionTracker.getRejectionReasons().entrySet().stream()
+                .map(entry -> NonBid.of(entry.getKey(), entry.getValue()))
+                .toList();
+
+        return SeatNonBid.of(bidder, nonBids);
+    }
+
+    private GreenbidsAdUnit createAdUnit(
+            Imp imp,
+            Map<String, Bid> seatsWithBids,
+            Map<String, NonBid> seatsWithNonBids) {
+        final ExtBanner extBanner = Optional.ofNullable(imp.getBanner())
+                .map(banner -> {
+                    if (banner.getFormat() == null) {
+                        throw new PreBidException("Error: Banner format should be non-null");
+                    }
+                    return ExtBanner.builder()
+                            .sizes(banner.getFormat())
+                            .pos(banner.getPos())
+                            .name(banner.getId())
+                            .build();
+                }).orElse(null);
+
+        final Video video = imp.getVideo();
+        final Native nativeObject = imp.getXNative();
+
+        final MediaTypes mediaTypes = MediaTypes.builder()
+                .banner(extBanner)
+                .video(video)
+                .nativeObject(nativeObject)
+                .build();
+
+        final List<GreenbidsBids> bids = extractBidders(imp, seatsWithBids, seatsWithNonBids);
+
+        final String adUnitCode = getAdUnitCode(imp);
```

Comment:
todo: Use the publisher provided adUnit code.
Use a dedicated field for unified code (`gpid | storedRequestId | imp.id`)

Reply:
fixed

As we have discussed internally I'have the class with nested fields `GreenbidsUnifiedCode` with the enum of code types.

```java
public class GreenbidsUnifiedCode {

    @JsonProperty("value")
    String value;

    @JsonProperty("src")
    Source source;

    public enum Source {
        gpidSource("gpid"),
        storedRequestIdSource("storedRequestId"),
        adUnitCodeSource("adUnitCode");
```

Also as discussed at the same time along with `GreenbidsUnifiedCode` we will keep logging `adUnitCode = imp.id` for the purpose of reporting